File: clojure/8-queens/8-queens.clj
(ns eight-queens
  (:import [javax.swing JPanel JFrame]
           [java.awt Graphics Color]))

(defn queens-intersect? [[x1 y1] [x2 y2]]
  (boolean (or (= x1 x2)
               (= y1 y2)
               (= (Math/abs (- x1 x2))
                  (Math/abs (- y1 y2))))))

(defn solve-queens
  ([dim] (solve-queens (into [] (for [x (range dim)] [x 0])) 0))
  ([queens curr-ind]
   (if (= (count queens) curr-ind)
     queens
     (loop [i (last (nth queens curr-ind))]
       (if (= i (count queens))
         nil
         (if (not-any? true? (map #(queens-intersect? % [curr-ind i]) (take curr-ind queens)))
           (let [res (solve-queens (assoc queens curr-ind (assoc (nth queens curr-ind) 1 i)) (inc curr-ind))]
             (if (not (nil? res))
               res
               (recur (inc i))))
           (recur (inc i))))))))



(defn make-panel [dim scl]
  (proxy [JPanel] []
    (paintComponent [#^Graphics g]
        (dorun (for [x (range dim) y (range dim)]
                 (do
                   (if (= (mod x 2) (mod y 2))
                       (. g setColor Color/WHITE)
                       (. g setColor Color/BLACK))
                   (. g fillRect (* x scl) (* y scl) scl scl))))
        (. g setColor Color/RED)
        (dorun (map #(. g fillOval (* (first %) scl) (* (last %) scl) scl scl)
                    (solve-queens dim))))))

(defn run-queens [dim]
  (let [scl 75]
    (doto (JFrame. "Queens!")
      (. setSize (* dim scl) (* dim scl))
      (.add (make-panel dim scl))
      (. setVisible true)
      (.setDefaultCloseOperation JFrame/DISPOSE_ON_CLOSE))))

(run-queens 5)

File: pythons/divesh/8_queens.py
def intersect(x1,y1,x2,y2):
    return x1 == x2 or y1 == y2 or abs(x1 - x2) == abs(y1 - y2)

def solve_queens_h(curr_sol,ind):
    if ind == len(curr_sol):
        return curr_sol
    while curr_sol[ind] < len(curr_sol):
        if not any(intersect(ind,curr_sol[ind],i,curr_sol[i]) for i in range(ind)):
            guess = solve_queens_h(curr_sol[:],ind + 1)
            if guess:
                return guess
        curr_sol[ind] = curr_sol[ind] + 1
    return False    

def solve_queens(dim):
    return solve_queens_h([0 for i in range(dim)],0)

def print_sol(sol):
    if not sol:
        print("No solution")
        return
    print('--')
    for i in sol:
        ln = ['#' for j in range(len(sol))]
        ln[i] = 'Q'
        print(''.join(ln))

if __name__ == "__main__":
    print_sol(solve_queens(8))
